File: src/chipschallenge/blockreactions/RandomForceFloorTo.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.BlockContainerFullException;
import chipschallenge.Inventory.Boots;
import chipschallenge.Move;
import chipschallenge.Move.Moves;

public class RandomForceFloorTo extends BlockReaction {

    private RandomForceFloorTo() {
    }
    private static RandomForceFloorTo mInstance = null;

    public static synchronized RandomForceFloorTo getInstance() {
        if (mInstance == null) {
            mInstance = new RandomForceFloorTo();
        }
        return mInstance;
    }

    @Override
    public void react(Block moving, Block standing) throws BlockContainerFullException {
        // Nothing
    }

    @Override
    public boolean canMove(Block moving, Block standing) {
        return moving.isChip();
    }

    @Override
    public Moves causesSlip(Block moving, Block standing) {
        if ((moving.isChip() && hasBoots(Boots.SUCTIONBOOTS))) {
            return null;
        } else {
            return Move.getRandom();
        }
    }
}

File: ChipsChallenge.Shared/Blockreactions/RandomForceFloorTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    using Boots = Inventory.Boots;
    using Moves = Move.Moves;

    public class RandomForceFloorTo : BlockReaction
    {
        private static RandomForceFloorTo instance;

        private RandomForceFloorTo()
        {
        }

        public static RandomForceFloorTo Instance
        {
            get
            {
                lock (typeof(RandomForceFloorTo))
                {
                    return instance ?? (instance = new RandomForceFloorTo());
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
        }

        public override bool canMove(Block moving, Block standing)
        {
            return moving.Chip;
        }

        public override Moves? CausesSlip(Block moving, Block standing)
        {
            if (moving.Chip && HasBoots(Boots.SUCTIONBOOTS))
            {
                return null;
            }

            return Move.Random;
        }
    }
}